### PR TITLE
Closes #102 | Add Dataloader for Thai SER

### LINKED ISSUE #102: Create dataset loader for THAI SER
Dataloader name: `thai_ser/thai_ser.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?thai_ser

|     Dataset| thai_ser |
|-------------|---|
| Description | THAI SER dataset consists of 5 main emotions assigned to actors: Neutral, Anger, Happiness, Sadness, and Frustration. The recordings were 41 hours, 36 minutes long (27,854 utterances), and were performed by 200 professional actors (112 female, 88 male) and directed by students, former alumni, and professors from the Faculty of Arts, Chulalongkorn University.  The THAI SER contains 100 recordings and is separated into two main categories: Studio and Zoom. Studio recordings also consist of two studio environments: Studio A, a controlled studio room with soundproof walls, and Studio B, a normal room without soundproof or noise control. |
| Subsets | - |
| Languages | tha |
| Tasks | Speech Emotion Recognition |
| License | Creative Commons Attribution Share Alike 4.0 (cc-by-sa-4.0) |
| Homepage | https://github.com/vistec-AI/dataset-releases/releases/tag/v1 |
| HF URL | - |
| Paper URL | https://airesearch.in.th/releases/speech-emotion-dataset/ |

### PR DESCRIPTION
**Pull request by zwenyu**
Closes #102 

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/{my_dataset}/{my_dataset}.py` (please use only lowercase and underscore for dataset folder naming, as mentioned in dataset issue) and its `__init__.py` within `{my_dataset}` folder.
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_LOCAL`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py` or `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py --subset_id {subset_name_without_source_or_seacrowd_suffix}`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

**Comment by holylovenia** [comment]
A friendly reminder for @tellarin and @fhudi to review.

File: seacrowd/sea_datasets/thai_ser/thai_ser.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            schema="source",
+            subset_id=f"{_DATASETNAME}_studio",
+        ),
+        SEACrowdConfig(
+            name=f"{_DATASETNAME}_studio_seacrowd_{SEACROWD_SCHEMA_NAME}",
+            version=SEACROWD_VERSION,
+            description=f"{_DATASETNAME} SEACrowd schema",
+            schema=f"seacrowd_{SEACROWD_SCHEMA_NAME}",
+            subset_id=f"{_DATASETNAME}_studio",
+        ),
```
Comment (fhudi) [inline seacrowd/sea_datasets/thai_ser/thai_ser.py:108]: I do not think that we should split this as subsets.
Based on the publication, it seems to be either with or without zoom
(https://github.com/vistec-AI/vistec-ser/blob/main/vistec_ser/examples/train_aisser.py#L17-L18)

---

IMHO, we have 3 options:
1. Follow the original splits of {studio, studio+zoom}
2. No splitting (single subset), as there is explicit indication of subsetting the dataset in the published writings.
3. Splits based on Recording Environments, i.e. {StudioA, StudioB, Zoom}, or even further splitting based on session types.

---

However, I am not an expert in this SER area nor able to speak Thai,
so please correct me if I'm wrong (CMIIW).
cc: @holylovenia

**Review by fhudi — CHANGES_REQUESTED** [review]
Hi @zwenyu, sorry for the delay in reviewing.
The seacrowd schema looks good to me, and it passed the tests.
However, I found it the source schema differs from the original.
Could you please make some changes? 🙏

File: seacrowd/sea_datasets/thai_ser/thai_ser.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                gen_kwargs={
+                    "actor_demography_filepath": data_paths["actor_demography"],
+                    "emotion_label_filepath": data_paths["emotion_label"],
+                    "data_filepath": data_paths[setting],
+                    "split": "train",
+                },
+            )
+        ]
+
+    def _generate_examples(self, actor_demography_filepath: Path, emotion_label_filepath: Path, data_filepath: Path, split: str) -> Tuple[int, Dict]:
```
Comment (fhudi) [inline seacrowd/sea_datasets/thai_ser/thai_ser.py:160]:
```suggestion
    def _generate_examples(self, actor_demography_filepath: Path, emotion_label_filepath: Path, data_filepath: Dict[str, Union[Path, Dict]], split: str) -> Tuple[int, Dict]:
```
The type-hint was wrong. This suggestion gives a more explicit (and more concise) declaration.
Please remember to import the `typing.Union` if the suggestion is adopted.

File: seacrowd/sea_datasets/thai_ser/thai_ser.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+                id = audio_path.split("/")[-1]
+                speaker_id = id.split("_")[2].strip("actor")
+                # labels in emotion_label are incomplete, labels only provided for microphone types: mic, con
+                # otherwise, obtain label from id for scripted utterances and skip sample for the improvised utterances
+                if id in emotion_label.keys():
+                    assigned_emo = emotion_label[id][0]["assigned_emo"]
+                else:
+                    if "script" in id:
+                        label = id.split("_")[-1][0]  # Emotion (1 = Neutral, 2 = Angry, 3 = Happy, 4 = Sad, 5 = Frustrated)
+                        assigned_emo = self._LABELS[int(label) - 1]
```
[the remark this answers is not in the document]
Comment (zwenyu) [reply seacrowd/sea_datasets/thai_ser/thai_ser.py:186]: For 'script' sessions, they are the same. For 'improvisation' sessions, the emotion label cannot be identified from `id`as there are two actors with different emotions.

File: seacrowd/sea_datasets/thai_ser/thai_ser.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+        # iterate through data folders
+        for folder_name, folder_path in data_filepath.items():
+            flac_files = glob.glob(os.path.join(folder_path, "**/*.flac"), recursive=True)
+            # iterate through recordings
+            for audio_path in flac_files:
+                id = audio_path.split("/")[-1]
+                speaker_id = id.split("_")[2].strip("actor")
+                # labels in emotion_label are incomplete, labels only provided for microphone types: mic, con
+                # otherwise, obtain label from id for scripted utterances and skip sample for the improvised utterances
```
[the remark this answers is not in the document]
Comment (zwenyu) [reply seacrowd/sea_datasets/thai_ser/thai_ser.py:180]: All IDs from `emotion_label.json'  have a FLAC file. There are a total of 61972 FLAC files, but only 27854 IDs.

File: seacrowd/sea_datasets/thai_ser/thai_ser.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        with open(actor_demography_filepath, "r", encoding="utf-8") as actor_demography_file:
+            actor_demography = json.load(actor_demography_file)
+        actor_demography_dict = {actor["Actor's ID"]: {"speaker_age": actor["Age"], "speaker_gender": actor["Sex"].lower()} for actor in actor_demography["data"]}
+
+        # read emotion_label file
+        with open(emotion_label_filepath, "r", encoding="utf-8") as emotion_label_file:
+            emotion_label = json.load(emotion_label_file)
+
+        # iterate through data folders
+        for folder_name, folder_path in data_filepath.items():
```
[the remark this answers is not in the document]
Comment (zwenyu) [reply seacrowd/sea_datasets/thai_ser/thai_ser.py]: Updated.

File: seacrowd/sea_datasets/thai_ser/thai_ser.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            flac_files = glob.glob(os.path.join(folder_path, "**/*.flac"), recursive=True)
+            # iterate through recordings
+            for audio_path in flac_files:
+                id = audio_path.split("/")[-1]
+                speaker_id = id.split("_")[2].strip("actor")
+                # labels in emotion_label are incomplete, labels only provided for microphone types: mic, con
+                # otherwise, obtain label from id for scripted utterances and skip sample for the improvised utterances
+                if id in emotion_label.keys():
+                    assigned_emo = emotion_label[id][0]["assigned_emo"]
```
[the remark this answers is not in the document]
Comment (zwenyu) [reply seacrowd/sea_datasets/thai_ser/thai_ser.py:198]: Added information in `source` schema and pushed changes.

File: seacrowd/sea_datasets/thai_ser/thai_ser.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+                if self.config.schema == "source" or self.config.schema == f"seacrowd_{self.SEACROWD_SCHEMA_NAME}":
+                    example = {
+                        "id": id.strip(".flac"),
+                        "path": audio_path,
+                        "audio": audio_path,
+                        "speaker_id": speaker_id,
+                        "labels": assigned_emo,
+                        "metadata": {"speaker_age": actor_demography_dict[speaker_id]["speaker_age"], "speaker_gender": actor_demography_dict[speaker_id]["speaker_gender"]},
+                    }
```
[the remark this answers is not in the document]
Comment (zwenyu) [reply seacrowd/sea_datasets/thai_ser/thai_ser.py]: Added information in source schema and pushed changes.

File: seacrowd/sea_datasets/thai_ser/thai_ser.py (hunk 1 above)
Comment (fhudi) [reply seacrowd/sea_datasets/thai_ser/thai_ser.py:108]: @zwenyu thanks for the updates on the other parts, LGTM.
As for this part, for the time being, could you please follow the split from the original?
i.e. normal and `--include-zoom`

Comment (zwenyu) [reply seacrowd/sea_datasets/thai_ser/thai_ser.py]: Updated to studio, and studio and zoom splits

**Review by fhudi — APPROVED** [review]
LGTM. thanks for the hardwork and cooperation, @zwenyu

**Review by patrickamadeus — APPROVED** [review]
Ran and passed the tests! So far there is no substantial changes needed from me apart from all previous improvements. LGTM!

**Comment by fhudi** [comment]
Merging this.
Thanks for the hardwork and contributions! 🙏

**Pull request merged**